go/vt/vtexplain/vtexplain.go

```diff
@@ -0,0 +1,207 @@
+	BindVars map[string]*querypb.BindVariable
+
+	// The actual queries executed by mysql
+	MysqlQueries []string
+}
+
+// MarshalJSON renders the json structure
+func (tq *TabletQuery) MarshalJSON() ([]byte, error) {
+	// Convert Bindvars to strings for nicer output
+	bindVars := make(map[string]string)
+	for k, v := range tq.BindVars {
+		bindVars[k] = v.String()
+	}
+
+	return json.Marshal(&struct {
+		SQL          string
+		BindVars     map[string]string
+		MysqlQueries []string
+	}{
+		SQL:          tq.SQL,
+		BindVars:     bindVars,
+		MysqlQueries: tq.MysqlQueries,
+	})
+}
+
+// Plan defines how vitess will execute a given sql query, including the vtgate
+// query plans and all queries run on each tablet.
+type Plan struct {
+	// original sql statement
+	SQL string
+
+	// the vtgate plan(s)
+	Plans []*engine.Plan
+
+	// list of queries / bind vars sent to each tablet
+	TabletQueries map[string][]*TabletQuery
+}
+
+const (
+	vtexplainCell = "explainCell"
+)
+
+// Init sets up the fake execution environment
+func Init(vSchemaStr, sqlSchema string, opts *Options) error {
+	// Verify options
+	if opts.ReplicationMode != "ROW" && opts.ReplicationMode != "STATEMENT" {
+		return fmt.Errorf("invalid replication mode \"%s\"", opts.ReplicationMode)
+	}
+
+	err := initVtgateExecutor(vSchemaStr, opts)
+	if err != nil {
+		return fmt.Errorf("initVtgateExecutor: %v", err)
+	}
+
+	parsedDDLs, err := parseSchema(sqlSchema)
+	if err != nil {
+		return fmt.Errorf("parseSchema: %v", err)
+	}
+
+	err = initTabletEnvironment(parsedDDLs, opts)
+	if err != nil {
+		return fmt.Errorf("initTabletEnvironment: %v", err)
+	}
+
+	return nil
+}
+
+func parseSchema(sqlSchema string) ([]*sqlparser.DDL, error) {
+	parsedDDLs := make([]*sqlparser.DDL, 0, 16)
+	for _, sql := range strings.Split(sqlSchema, ";") {
+		s := sqlparser.StripLeadingComments(sql)
+		s, _ = sqlparser.SplitTrailingComments(sql)
+		s = strings.TrimSpace(s)
+		if s == "" {
+			continue
+		}
+
+		stmt, err := sqlparser.Parse(sql)
+		if err != nil {
+			log.Errorf("ERROR: failed to parse sql: %s, got error: %v", sql, err)
+			continue
+		}
+		ddl, ok := stmt.(*sqlparser.DDL)
+		if !ok {
+			log.Infof("ignoring non-DDL statement: %s", sql)
+			continue
+		}
+		if ddl.Action != sqlparser.CreateStr {
+			log.Infof("ignoring %s table statement", ddl.Action)
+			continue
+		}
+		if ddl.TableSpec == nil {
+			log.Errorf("invalid create table statement: %s", sql)
+			continue
+		}
+		parsedDDLs = append(parsedDDLs, ddl)
+	}
+	return parsedDDLs, nil
+}
+
+// Run the explain analysis on the given queries
+func Run(sqlStr string) ([]*Plan, error) {
+	plans := make([]*Plan, 0, 16)
+
+	for _, sql := range strings.Split(sqlStr, ";") {
+		s := strings.TrimSpace(sql)
+		if s != "" {
+			plan, err := getPlan(s)
+			if err != nil {
+				return nil, err
+			}
+			plans = append(plans, plan)
+		}
+	}
+
+	return plans, nil
+}
+
+func getPlan(sql string) (*Plan, error) {
+	_, err := sqlparser.Parse(sql)
```

Comment:
Will it be better to send the query through and let vtgate respond? The error will look more like what the app would see, although less readable. On the fence.